File: Seguimiento/Seguimiento1/N1152222545/Runge-Kutta-6.py
# -*- coding: utf-8 -*-
"""
Created on Wed Sep 18 20:53:04 2019

@author: Samuel
"""

import numpy as np
import matplotlib.pyplot as plt 
import pandas as pd
import os 


os.system("g++ -o RK6 Runge-Kutta-6.cpp && ./RK6")

datos = pd.read_csv("RK6.csv", sep=" ")

X=datos['X']
Y=datos['Y']
YE=datos['YE']
Diff=datos['Diff']


xf=5.
x0=0

Num=np.array([10,100,1000,10000])
h=(xf-x0)/Num

ax0=plt.subplot(131)
ax1=plt.subplot(132)
ax2=plt.subplot(133)


#se grafica los resultados para distintos numeros de puntos, lo que cambia también el paso 
ax0.plot(X.iloc[112:1113],YE.iloc[112:1113],'r',label= 'Exacta')
ax0.plot(X.iloc[0:11],Y.iloc[0:11],'g--',label= 'Nump=10')
ax0.plot(X.iloc[11:112],Y.iloc[11:112],'b--',label='Nump=100')
ax0.plot(X.iloc[112:1113],Y.iloc[112:1113],'c--',label= 'Nump=1000')
ax0.plot(X.iloc[1113:11114],Y.iloc[1113:11114],'m',label= 'Nump=10000')
ax0.set_ylabel("y")
ax0.set_xlabel("x")

ax0.legend()


ax1.semilogy(X[0:11],Diff[0:11],'r--',label= 'Nump=10-err')
ax1.semilogy(X[11:112],Diff[11:112],'b--',label= 'Nump=100-err')
ax1.semilogy(X[112:1113],Diff[112:1113],'c--',label= 'Nump=1000-err')
ax1.semilogy(X[1113:11114],Diff[1113:11114],'m--',label= 'Nump=10000-err')
ax1.set_ylabel("Error")
ax1.set_xlabel("x")

ax1.legend()

DifP=[]

DifP.append(np.mean(Diff[0:11]))
DifP.append(np.mean(Diff[11:112]))
DifP.append(np.mean(Diff[112:1113]))
DifP.append(np.mean(Diff[1113:11114]))

ax2.plot(h,DifP,label='Error vs Paso')
ax2.set_xlabel("h")
ax2.set_ylabel("Prom-err")
ax2.legend()

plt.show()

File: Seguimiento/Seguimiento1/N1152222545/Runge-Kutta-6.cpp
#include <iostream>
#include <math.h>
#include <fstream>

using namespace std;



float F(float xn,float yn)
  {
    return xn-yn;

  }
float f(float x, float y)
{
  return x-1+2*exp(-x);
}
  

float RK6(float xn,float yn,float h)

{

 float k1 = F(xn,yn);
  float k2 = F(xn + h/3.,yn + h/3.*k1);
  float k3 = F(xn + h*3./4.,yn + h*((123./256.)*k1 + (315./256.)*k2));
  float k4 = F(xn + h/5.,yn + h*((193./750.)*k1 + (189./1250.)*k2 + (176./1875.)*k3));
  float k5 = F(xn + h*2/3,yn + h*((26./81.)*k1 + (7./15.)*k2 + (304./4455.)*k3 + (175/297)*k4));
  float k6 = F(xn + h, yn + h*((151./150.)*k1 + (351./250.)*k2 + (304./4125.)*k3 + (5/77)*k4 + (243./175.)*k5));
 return yn + h*(k1/24 + (125./336.)*k4 + (27./56.)*k5 + (5./48.)*k6);			     
			       
}

int main()

{
 float y0=1.0;
  float x0=0.0;
  float xf=5.0;
  int Num[4] = {10,100,1000,10000};
  int d=sizeof(Num)/sizeof(*Num);
  float h[d];

  ofstream ofs ("RK6.csv",ofstream::out);


  ofs<<"X"<<" "<<"Y"<<" "<<"YE"<<" "<<"Diff"<<endl;
for (int j =0; j<d;j++)
    {
  h[j]= (xf-x0)/Num[j];
   float X[Num[j]]={x0};
  float Y[Num[j]]={y0};
  float Diff[Num[j]]={0};


  ofs << X[0]<< " "<< Y[0]<<" "<<X[0]-1+2*exp(-X[0])<<" "<<Diff[0]<<endl;

for (int i = 0; i<Num[j]; i++)
    {

        X[i+1]=X[i]+ h[j];
      Y[i+1]=RK6(X[i],Y[i],h[j]);
      Diff[i+1]=fdim(Y[i+1], X[i]-1+2*exp(-X[i]));

      ofs << X[i+1]<< " "<< Y[i+1]<<" "<<X[i]-1+2*exp(-X[i])<<" "<<Diff[i+1]<<endl;


      
	//cout<<X[i+1]<< " "<<Y[i+1]<<endl;
      
	
     
    }


    }


 ofs.close();
   
 return 0;
}
